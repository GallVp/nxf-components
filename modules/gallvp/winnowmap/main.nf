process WINNOWMAP {
    tag "${meta.id}"
    label 'process_high'

    container "ghcr.io/gallvp/winnowmap:sha-bb3c814"

    input:
    tuple val(meta), path(reads)
    tuple val(meta2), path(reference)
    tuple val(meta3), path(merylDB)
    val meryl_distinct
    val bam_format
    val bam_index_extension
    val cigar_paf_format
    val cigar_bam
    val sort_bam

    output:
    tuple val(meta), path("*.paf"), optional: true, emit: paf
    tuple val(meta), path("*.bam"), optional: true, emit: bam
    tuple val(meta), path("*.bam.${bam_index_extension}"), optional: true, emit: index
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def bam_index = bam_index_extension ? "${prefix}.bam##idx##${prefix}.bam.${bam_index_extension} --write-index" : "${prefix}.bam"
    def samtools_command = sort_bam ? 'sort' : 'view'
    def bam_output = bam_format ? "-a | samtools ${samtools_command} -@ ${task.cpus - 1} -o ${bam_index} ${args2}" : "-o ${prefix}.paf"
    def cigar_paf = cigar_paf_format && !bam_format ? "-c" : ''
    def set_cigar_bam = cigar_bam && bam_format ? "-L" : ''
    def bam_input = "${reads.extension}".matches('sam|bam|cram')
    def samtools_reset_fastq = bam_input ? "samtools fastq --threads ${task.cpus - 1} ${args3} ${reads} > ${prefix}.reset.fastq" : ''
    def query = bam_input ? "${prefix}.reset.fastq" : reads
    def target = reference ?: (bam_input ? error("BAM input requires reference") : reads)

    """
    meryl \\
        print \\
        greater-than \\
        distinct=${meryl_distinct}\\
        ${merylDB} > repetitive_kx.txt

    ${samtools_reset_fastq}
    winnowmap \\
        ${args} \\
        -W repetitive_kx.txt \\
        -t ${task.cpus} \\
        ${target} \\
        ${query} \\
        ${cigar_paf} \\
        ${set_cigar_bam} \\
        ${bam_output}


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        winnowmap: \$(winnowmap --version 2>&1)
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def output_file = bam_format ? "${prefix}.bam" : "${prefix}.paf"
    def bam_index = bam_index_extension ? "touch ${prefix}.bam.${bam_index_extension}" : ""
    def bam_input = "${reads.extension}".matches('sam|bam|cram')
    def target = reference ?: (bam_input ? error("BAM input requires reference") : reads)

    """
    touch ${output_file}
    ${bam_index}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        winnowmap: \$(winnowmap --version 2>&1)
    END_VERSIONS
    """
}
