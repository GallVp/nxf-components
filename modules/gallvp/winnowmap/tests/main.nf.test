nextflow_process {

    name "Test Process WINNOWMAP"
    script "../main.nf"
    process "WINNOWMAP"

    tag "modules"
    tag "modules_gallvp"
    tag "winnowmap"
    tag "meryl/count"

    setup {
        run("MERYL_COUNT") {
            script "../../meryl/count/main.nf"
            process {
                """
                input[0] = [
                    [id: 'test'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = 15
                """
            }
        }
    }

    test("sarscov2 - fastq, fasta, true, [], false, false") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = MERYL_COUNT.out.meryl_db
                input[3] = 0.9998
                input[4] = true
                input[5] = []
                input[6] = false
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam[0][1]).getHeader(),
                    bam(process.out.bam[0][1]).getReadsMD5(),
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - fastq, fasta, true, 'bai', false, false") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = MERYL_COUNT.out.meryl_db
                input[3] = 0.9998
                input[4] = true
                input[5] = 'bai'
                input[6] = false
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam[0][1]).getHeader(),
                    bam(process.out.bam[0][1]).getReadsMD5(),
                    file(process.out.index[0][1]).name,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - [fastq1, fastq2], fasta, true, false, false") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = MERYL_COUNT.out.meryl_db
                input[3] = 0.9998
                input[4] = true
                input[5] = []
                input[6] = false
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam[0][1]).getHeader(),
                    bam(process.out.bam[0][1]).getReadsMD5(),
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - bam, fasta, true, [], false, false") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test3.single_end.markduplicates.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = MERYL_COUNT.out.meryl_db
                input[3] = 0.9998
                input[4] = true
                input[5] = []
                input[6] = false
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam[0][1]).getHeader(),
                    bam(process.out.bam[0][1]).getReadsMD5(),
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - bam, fasta, true, 'bai', false, false") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test3.single_end.markduplicates.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = MERYL_COUNT.out.meryl_db
                input[3] = 0.9998
                input[4] = true
                input[5] = 'bai'
                input[6] = false
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam[0][1]).getHeader(),
                    bam(process.out.bam[0][1]).getReadsMD5(),
                    file(process.out.index[0][1]).name,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - fastq, fasta, true, [], false, false - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = MERYL_COUNT.out.meryl_db
                input[3] = 0.9998
                input[4] = true
                input[5] = []
                input[6] = false
                input[7] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}