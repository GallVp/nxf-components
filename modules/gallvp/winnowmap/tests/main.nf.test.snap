{
    "sarscov2 - bam, fasta, true, 'bai', false, false": {
        "content": [
            [
                "@HD\tVN:1.6\tSO:coordinate",
                "@SQ\tSN:MT192765.1\tLN:29829",
                "@PG\tID:Winnowmap\tPN:Winnowmap\tVN:2.03\tCL:winnowmap -W repetitive_kx.txt -t 2 -a genome.fasta test.reset.fastq",
                "@PG\tID:samtools\tPN:samtools\tPP:Winnowmap\tVN:1.16.1\tCL:samtools sort -@ 1 -o test.bam##idx##test.bam.bai --write-index"
            ],
            "1f769ca2863ec9d9c79f75142c7fe58c",
            "test.bam.bai",
            [
                "versions.yml:md5,44e1c02faad26af0d0bdb30844faeff6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-28T14:27:29.774593"
    },
    "sarscov2 - fastq, fasta, true, [], false, false - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,9b1226a11789e59d2e1dead92de69995"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "index": [
                    
                ],
                "paf": [
                    
                ],
                "versions": [
                    "versions.yml:md5,9b1226a11789e59d2e1dead92de69995"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-28T14:29:26.218865"
    },
    "sarscov2 - [fastq1, fastq2], fasta, true, false, false": {
        "content": [
            [
                "@HD\tVN:1.6\tSO:coordinate",
                "@SQ\tSN:MT192765.1\tLN:29829",
                "@PG\tID:Winnowmap\tPN:Winnowmap\tVN:2.03\tCL:winnowmap -W repetitive_kx.txt -t 2 -a genome.fasta test_1.fastq.gz test_2.fastq.gz",
                "@PG\tID:samtools\tPN:samtools\tPP:Winnowmap\tVN:1.16.1\tCL:samtools sort -@ 1 -o test.bam"
            ],
            "6c1dd20d64c8f6b5aad6d4c5a88cfeb5",
            [
                "versions.yml:md5,44e1c02faad26af0d0bdb30844faeff6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-28T14:24:29.290371"
    },
    "sarscov2 - fastq, fasta, true, [], false, false": {
        "content": [
            [
                "@HD\tVN:1.6\tSO:coordinate",
                "@SQ\tSN:MT192765.1\tLN:29829",
                "@PG\tID:Winnowmap\tPN:Winnowmap\tVN:2.03\tCL:winnowmap -W repetitive_kx.txt -t 2 -a genome.fasta test_1.fastq.gz",
                "@PG\tID:samtools\tPN:samtools\tPP:Winnowmap\tVN:1.16.1\tCL:samtools sort -@ 1 -o test.bam"
            ],
            "f4019e8862cadae86beb29fff8c28c85",
            [
                "versions.yml:md5,44e1c02faad26af0d0bdb30844faeff6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-28T14:20:36.037233"
    },
    "sarscov2 - fastq, fasta, true, 'bai', false, false": {
        "content": [
            [
                "@HD\tVN:1.6\tSO:coordinate",
                "@SQ\tSN:MT192765.1\tLN:29829",
                "@PG\tID:Winnowmap\tPN:Winnowmap\tVN:2.03\tCL:winnowmap -W repetitive_kx.txt -t 2 -a genome.fasta test_1.fastq.gz",
                "@PG\tID:samtools\tPN:samtools\tPP:Winnowmap\tVN:1.16.1\tCL:samtools sort -@ 1 -o test.bam##idx##test.bam.bai --write-index"
            ],
            "f4019e8862cadae86beb29fff8c28c85",
            "test.bam.bai",
            [
                "versions.yml:md5,44e1c02faad26af0d0bdb30844faeff6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-28T14:24:23.458085"
    },
    "sarscov2 - bam, fasta, true, [], false, false": {
        "content": [
            [
                "@HD\tVN:1.6\tSO:coordinate",
                "@SQ\tSN:MT192765.1\tLN:29829",
                "@PG\tID:Winnowmap\tPN:Winnowmap\tVN:2.03\tCL:winnowmap -W repetitive_kx.txt -t 2 -a genome.fasta test.reset.fastq",
                "@PG\tID:samtools\tPN:samtools\tPP:Winnowmap\tVN:1.16.1\tCL:samtools sort -@ 1 -o test.bam"
            ],
            "1f769ca2863ec9d9c79f75142c7fe58c",
            [
                "versions.yml:md5,44e1c02faad26af0d0bdb30844faeff6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-28T14:27:24.195849"
    }
}